File: src/types/values.ts
// Primitive types that can be stored in collection types.
// Includes JSON-serialisable data so that maps and lists
// can hold plain JS values.
export type Primitive =
  | string
  | number
  | boolean
  | Buffer
  // JSON-serialisable primitive values
  | { [key: string]: Primitive }
  | Primitive[];

// Unique symbol for nominal typing within TypeScript's structural type system.
// This prevents structural compatibility between LiveObject types.
declare const __livetype: unique symbol;

// Branded interfaces that enables TypeScript to distinguish
// between LiveObject types even when they have identical structure.
// Enables PathObject<T> to dispatch to correct method sets via conditional types.

export interface LiveMap<
  T extends Record<string, Value> = Record<string, Value>,
> {
  [__livetype]: "LiveMap";
}

export interface LiveList<T extends Value = Value> {
  [__livetype]: "LiveList";
}

export interface LiveCounter {
  [__livetype]: "LiveCounter";
}

// Type union that matches any LiveObject type that can be mutated, subscribed to, etc.
export type LiveObject = LiveMap | LiveList | LiveCounter;

// Type union that defines the base set of allowed types that can be stored in collection types.
// Describes the set of all possible values that can parameterize PathObject.
// This is the canonical union used when we cannot infer a narrower type.
export type Value = LiveObject | Primitive;

// LiveObject type class implementations.
// Currently only defines the static method used to create branded value types
// which can be provided as an argument to mutation methods.

export class LiveMap {
  static create<T extends Record<string, Value>>(
    initialData?: T,
  ): LiveMap<T extends Record<string, Value> ? T : {}> {
    return undefined as any;
  }
}

export class LiveList {
  static create<T extends Value>(
    initialData?: T[],
  ): LiveList<T extends Value ? T : Value> {
    return undefined as any;
  }
}

export class LiveCounter {
  static create(initialValue: number = 0): LiveCounter {
    return undefined as any;
  }
}

File: swift/Sources/AblyLiveObjects/values.swift
import Foundation

// Lawrence: in our previous LiveObjects-swift we ditched this type and incorporated it into LiveMapValue, see if that is something to do again. It's just that "Primitive" is mentioned a lot more as part of the public API now (it's in PathObject)
// Primitive types that can be stored in collection types.
// Includes JSON-serialisable data so that maps and lists
// can hold plain JS values.
public enum Primitive {
    case string(String)
    case number(Double)
    case bool(Bool)
    case data(Data)
    case jsonArray([JSONValue])
    case jsonObject([String: JSONValue])
}

// Type union that defines the base set of allowed types that can be stored in collection types.
// Describes the set of all possible values that can parameterize PathObject.
// This is the canonical union used when we cannot infer a narrower type.
public enum Value {
    // Lawrence: This is like the LiveMapValue type in the current LiveObjects plugin (have reverted to using the `primitive` naming given by TS, will revisit)
    case primitive(Primitive)
    case liveMap(LiveMap)
    case liveCounter(LiveCounter)
    case liveList(LiveList)
}

// LiveObject type class implementations.
// Currently only defines the static method used to create branded value types
// which can be provided as an argument to mutation methods.

// swiftformat:disable enumNamespaces
public struct LiveMap {
    public static func create(
        initialData _: [String: Value] = [:],
    ) -> Self {
        fatalError("Not implemented")
    }
}

public struct LiveList {
    public static func create(
        initialData _: [Value] = [],
    ) -> Self {
        fatalError("Not implemented")
    }
}

public struct LiveCounter {
    public static func create(
        initialValue _: Double = 0,
    ) -> Self {
        fatalError("Not implemented")
    }
}

// swiftformat:enable enumNamespaces
